### PR TITLE
add relative path support for creation rule

### PR DESCRIPTION
**Pull request by hamidb80**
addresses https://github.com/suchnsuch/Tangent/issues/452

### CODE REVIEW

File: apps/tangent-electron/src/app/model/commands/CreateNewFile.ts [hunk 1]
```diff
@@ -117,7 +117,21 @@ export default class CreateNewFileCommand extends WorkspaceCommand {
 			}
 
 			// Combining into a relative path allows rule name templates to define folders
-			relativePath = paths.join(folderPath || rule.folder, name + '.md')
+			let relativeDirPath = folderPath || rule.folder
+			if (relativeDirPath.startsWith('.')) {
+				const workspaceDir = this.workspace.viewState.directoryView.root.path
+				const viewState = this.workspace.viewState.tangent.getCurrentViewState()
+				const absoluteParentDir = viewState 
+								? (viewState.node.fileType == 'folder')
```
Comment (taylorhadden) [inline apps/tangent-electron/src/app/model/commands/CreateNewFile.ts]: Nit: this is a very large amount of indentation here. I recommend a single additional tab from the current scope.

File: apps/tangent-electron/src/app/model/commands/CreateNewFile.ts [hunk 2]
```diff
@@ -117,7 +117,21 @@ export default class CreateNewFileCommand extends WorkspaceCommand {
 			}
 
 			// Combining into a relative path allows rule name templates to define folders
-			relativePath = paths.join(folderPath || rule.folder, name + '.md')
+			let relativeDirPath = folderPath || rule.folder
+			if (relativeDirPath.startsWith('.')) {
+				const workspaceDir = this.workspace.viewState.directoryView.root.path
+				const viewState = this.workspace.viewState.tangent.getCurrentViewState()
```
Comment (taylorhadden) [inline apps/tangent-electron/src/app/model/commands/CreateNewFile.ts]: You could also do `...getCurrentViewState()?.node` as you only ever reference `node` in the view state.

File: apps/tangent-electron/src/app/model/commands/CreateNewFile.ts [hunk 3]
```diff
@@ -117,7 +117,21 @@ export default class CreateNewFileCommand extends WorkspaceCommand {
...
+				const absoluteParentDir = viewState 
+								? (viewState.node.fileType == 'folder') 
+									? viewState.node.path                // in folder
+									: paths.dirname(viewState.node.path) // in file
+								:  workspaceDir                          // at root level
+				const absoluteParentPath = paths.join(absoluteParentDir, relativeDirPath)
+				relativeDirPath = paths
+								.resolve(absoluteParentPath) // resolve . or ..
+				    			.substring(workspaceDir.length) // make path relative to the workspace
```
Comment (taylorhadden) [inline apps/tangent-electron/src/app/model/commands/CreateNewFile.ts]: While this should work, I recommend using `workspace.directoryStore.pathToRelativePath()`. Alternatively, since you know the root already, you can use `paths.getChildPath()`.

**Review by taylorhadden — CHANGES_REQUESTED** [review]
Looks pretty solid overall. None of my comments are big issues.

I would love to see some tests added to `CreateNewFile.test.ts`!

File: apps/tangent-electron/src/app/model/commands/CreateNewFile.ts (hunk 2 above)
Comment (hamidb80) [reply apps/tangent-electron/src/app/model/commands/CreateNewFile.ts]: no I've used `viewState.node` 3 times in the code, removing the `viewState` makes it messy

**Comment by hamidb80** [comment]
Hi @taylorhadden . sorry for interruption.

seems like `workspace.viewState.tangent` is not available at test time!?!. what should I do?


```
TypeError: Cannot read properties of undefined (reading 'getCurrentViewState')
 ❯ CreateNewFileCommand.resolveContext src/app/model/commands/CreateNewFile.ts:122:50
    122|     const node = this.workspace.viewState.tangent.getCurrentViewState(…
       |                                                  ^
```

**Comment by taylorhadden** [comment]
Looking at this again, does this actually handle the case described in #452? To resolve a path with a relative section, we need three parts: the base, the relative folder, and the name section. It seems like the base is replacing the relative section.

**Comment by hamidb80** [comment]
> does this actually handle the case described in #452? 

Yes I've tested it manually. It works in these cases:
1. Root (no file opened)
2. In a directory
3. In a file

The test path was `./` and the file name was `temp`

**Comment by hamidb80** [comment]
It may need more tests but I first need to get around https://github.com/suchnsuch/Tangent/pull/454#issuecomment-5691108905

**Comment by taylorhadden** [comment]
Sorry, I missed that. At the top of the file for `CreatenewFile.test.ts`, there is a `const workspace` definition that contains a bunch of dummy mock objects. You can either choose to implement a mock that gets you what you need for the test, or you can configure the code to handle a null `.tangent` field.

What happens if the test path is something other than `./`? For instance `../MyFolder` or `./OtherFolder`? We can't allow relative file paths to escape out of the workspace.

**Comment by hamidb80** [comment]
> What happens if the test path is something other than `./`? For instance `../MyFolder` or `./OtherFolder`? We can't allow relative file paths to escape out of the workspace.

I've added some tests. the path would not jump out of workspace

File: apps/tangent-electron/src/app/model/commands/CreateNewFile.test.ts [hunk 4]
```diff
@@ -3,39 +3,46 @@ import CreateNewFileCommand, { type CreateNewFileCommandContext } from './Create
...
-
-	const workspace = {
-		directoryStore,
-		viewState: {
-			directoryView: {
-				selection: {
-					value: []
+	// Expose the private function type-safely
+	function resolveContext(context: CreateNewFileCommandContext, node?: TreeNode) {
+		const directoryStore = new IndexTreeStore({
```
Comment (taylorhadden) [inline apps/tangent-electron/src/app/model/commands/CreateNewFile.test.ts:11]: FWIW, this `directoryStore` and `workspace` is never changed and can be a const value outside of the function so that it doesn't get reconstructed on every use.

File: apps/tangent-electron/src/app/model/commands/CreateNewFile.test.ts [hunk 5]
```diff
@@ -109,4 +116,85 @@ describe('Extension auto inclusion', () => {
...
+				mode: 'createOrOpen',
+				contentTemplate: undefined,
+				description: undefined
+			},
+			extension: 'default-md'
+		}, {
+				path: '/some/root/note.md',
+				name: 'test note',
+				fileType: 'note',
+				createShallowCopy(){ return {} as TreeNode }
```
Comment (taylorhadden) [inline apps/tangent-electron/src/app/model/commands/CreateNewFile.test.ts:135]: Is this stub actually necessary?

File: apps/tangent-electron/src/app/model/commands/CreateNewFile.test.ts [hunk 6]
```diff
@@ -109,4 +116,132 @@ describe('Extension auto inclusion', () => {
 			creationMode: undefined
 		})
 	})
+
+	it('relative path in note: ./', () => {
```
Comment (taylorhadden) [inline apps/tangent-electron/src/app/model/commands/CreateNewFile.test.ts:120]: Minor style note: the convention is to write the name of the test as if it was a descriptive sentence starting with "it". For instance, here might be "(it) Resolves paths adjacent to a note" or something. The name of the test should broadly assert what the expectation is.

File: apps/tangent-electron/src/app/model/commands/CreateNewFile.ts [hunk 7]
```diff
@@ -117,7 +117,18 @@ export default class CreateNewFileCommand extends WorkspaceCommand {
...
+			let relativeDirPath = folderPath || rule.folder
+			if (relativeDirPath.startsWith('.')) {
+				const node = this.workspace.viewState.tangent.getCurrentViewState()?.node
+				const relativeNodePath = node && 
+					this.workspace.directoryStore.pathToRelativePath(
+						(node.fileType == 'folder') 
+							? node.path                // in folder
+							: paths.dirname(node.path) // in file
+					) || '' // at root level
+				relativeDirPath = paths.resolve(relativeNodePath + '/' + relativeDirPath)
```
Comment (taylorhadden) [inline apps/tangent-electron/src/app/model/commands/CreateNewFile.ts:129]: Don't manually insert path separators. Always use `paths.join()`.

File: apps/tangent-electron/src/app/model/commands/CreateNewFile.ts [hunk 8]
```diff
@@ -117,7 +117,18 @@ export default class CreateNewFileCommand extends WorkspaceCommand {
...
+			}
+			relativePath = paths.join(relativeDirPath, name + '.md')
```
Comment (taylorhadden) [inline apps/tangent-electron/src/app/model/commands/CreateNewFile.ts:131]: If the named template also includes a relative path, this will break here. Perhaps worth fixing with a `resolve()`?

**Review by taylorhadden — CHANGES_REQUESTED** [review]
Looking good! Just some relatively minor notes besides the manual slash inclusion.

File: apps/tangent-electron/src/app/model/commands/CreateNewFile.test.ts (hunk 5 above)
Comment (hamidb80) [reply apps/tangent-electron/src/app/model/commands/CreateNewFile.test.ts]: I'll remove `createShallowCopy`

File: apps/tangent-electron/src/app/model/commands/CreateNewFile.ts (hunk 8 above)
Comment (hamidb80) [reply apps/tangent-electron/src/app/model/commands/CreateNewFile.ts]: I didn't change this line. the behavior is as before.

maybe a test could reveal the bug?

Comment (taylorhadden) [reply apps/tangent-electron/src/app/model/commands/CreateNewFile.ts]: Sure, but since relative path support is being added for the folder, it makes sense to do so from the name template as well, no?
